src/bridgetree.rs

```diff
@@ -584,32 +612,109 @@ impl<H: Hashable + Hash + Eq + Debug, const DEPTH: u8> Debug for BridgeTree<H, D
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
         write!(
             f,
-            "BridgeTree {{\n  depth: {:?},\n  bridges: {:?},\n  incomplete_from: {:?},\n  saved: {:?},\n  checkpoints: {:?},\n  max_checkpoints: {:?}\n}}",
-            DEPTH, self.bridges, self.incomplete_from, self.saved, self.checkpoints, self.max_checkpoints
+            "BridgeTree {{\n  depth: {:?},\n  bridges: {:?},\n saved: {:?},\n  checkpoints: {:?},\n  max_checkpoints: {:?}\n}}",
+            DEPTH, self.bridges, self.saved, self.checkpoints, self.max_checkpoints
         )
     }
 }
 
+#[derive(Debug, Clone)]
+pub enum BridgeTreeError {
+    IncorrectIncompleteIndex,
+    InvalidWitnessIndex,
+    InvalidSavePoints,
+    ContinuityError,
+    CheckpointMismatch,
+}
+
+impl<H: Hash + Eq, const DEPTH: u8> BridgeTree<H, DEPTH> {
+    /// Removes the oldest checkpoint. Returns true if successful and false if
+    /// there are no checkpoints.
+    fn drop_oldest_checkpoint(&mut self) -> bool {
+        if self.checkpoints.is_empty() {
+            false
+        } else {
+            self.checkpoints.remove(0);
+            true
+        }
+    }
+
+    pub fn bridges(&self) -> &Vec<MerkleBridge<H>> {
+        &self.bridges
+    }
+
+    pub fn witnessable_leaves(&self) -> &HashMap<H, usize> {
+        &self.saved
+    }
+
+    pub fn checkpoints(&self) -> &Vec<Checkpoint<H>> {
```

Comment:
```suggestion
    pub fn checkpoints(&self) -> &[Checkpoint<H>] {
```
All these new public methods also need documentation.